bite-size-notes/week-5-c-Reachability-Analysis.md

```diff
@@ -0,0 +1,215 @@
+
+We may also care about entering Target set not on final time but any time between [0,T]
+## Hamilton-Jacobi Reachability
+Hamilton-Jacobi Reachability is a specific method used in the context of reachability analysis. It uses the Hamilton-Jacobi-Bellman (HJB) or Hamilton-Jacobi- Isaacs (HJI) equations, which are partial differential equations that describe the evolution of the reachability boundary with time. 
+
+There is a HJ toolbox written in in stanford ASL written for python, and matlab available, and can be imported with the command: import hj_reachability in phython. It is a toolbow that allows us to solve reachability problems.
+
+-Example From HJ toolbox
+
+-Example System:Air3d  (Pursuer trying to reach to an Evader)
+
+-System Dynamics:
+
+$$
+\begin{aligned}
+\dot{x} & = v \cos(\theta), \\
+\dot{y} & = v \sin(\theta), \\
+\dot{\theta} & = \omega.
+\end{aligned}
+$$
+
+### Code:
+
+```python
+import hj # Assuming 'hj' is a hypothetical module for Hamilton-Jacobi analysis
+
+# Initialize the dynamics for a 3D air system
+dynamics = hj.systems.Air3d()
+
+# Create a grid with specific boundary conditions and lattice parameters
+grid = hj.Grid.from_lattice_parameters_and_boundary_conditions(
+    hj.sets.Box(np.array([-6., -10., 0.]), np.array([20., 10., 2 * np.pi])),
+    (51, 40, 50),
+    periodic_dims=2
+)
+
+# Compute the values as the norm of states, adjusted by a radius of 5 units
+values = jnp.linalg.norm(grid.states[..., :2], axis=-1) - 5
+
+# Define solver settings with very high accuracy and a specific postprocessor for reachability
+solver_settings = hj.SolverSettings.with_accuracy(
+    "very_high",
+    hamiltonian_postprocessor=hj.solver.backwards_reachable_tube
+)
+time = 0.
+target_time = -2.8
+target_values = hj.step(solver_settings, dynamics, grid, time, values, target_time)
+```
+
+-"dynamics" line helps setting up dynamics
+
+-"grid"  line specifiies mesh, the grid points we are solving the ODE over
+
+-"vaules" line specifies the terminal point (we need a termenial point to sole ODE backwards in time. And specifies the minimum distance allowable between 2 planes, which is 5, so we form a circle of 5.
+
+-"solver_setting" we set setting with very high precision, solving for the backward reachable tube.
+
+-"target_time" sets that we are gonna go backwards in time for 2.8 seconds.
+
+-"target_values" solver line.
+
+-Solution Set
+<img width="906" alt="Screen Shot 2024-04-29 at 22 32 19" src="https://github.com/kutaydemiralay/AA548-spr2024/assets/160373451/00e43353-7caf-4f4e-9e88-bc8990905ce0">
+
+Contours values show that within 2.8 seconds if the pursuer is in the negative contour set, it would catch the evader. When the pursuer is out of the negative set, within 2.8 seconds, it cannot reach the evader, no matter what. Considering (0,0) is your referance frame and you are the evader.
+## Preliminaries
```

Comment:
I think it would make it easier for the reader if you could write the preliminaries and math, first and then write the code and example.

Reply:
also, I think you can use bullet lists/ numbered lists to make it easier to read:
 Markup : * Bullet list
              * Nested bullet
                  * Sub-nested bullet etc
          * Bullet list item 2

-OR-

 Markup : - Bullet list
              - Nested bullet
                  - Sub-nested bullet etc
          - Bullet list item 2 
          
Markup : 1. A numbered list
                        1. A nested numbered list
                        2. Which is numbered
               2. Which is numbered